doc/man/man4/m3ua.4.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man4/m3ua.4.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH M3UA 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.\"
.\"
.SH NAME
.B m3ua
\- SS7 MTP3 User Adaptation Layer (M3UA)
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <sys/mtpi.h>
.br
.B #include <sys/mtpi_ioctl.h>
.br
.B #include <sys/m3ua_ioctl.h>
.HP
.BI "int " m3ua_stream " = open(" device ", " flags );
.HP
.BI "int " m3ua_xtistr " = t_open(" device ", " flags );
.HP
.BI "int " m3ua_socket " = socket(" \fRPF_SS7\fI ", " socket_type ", " \fRSS7PROTO_M3UA\fI );
.\"
.\"
.SH DESCRIPTION
.PP
.B M3UA
is an implementation of the SS7 MTP3 User Adaptation Layer (M3UA) defined in
RFC 3332,
.[
rfc3332
.]
TS 102 142,
.[
ts 102 142
.]
and
RFC 4666.
.[
rfc4666
.]
.B M3UA
provides a back haul of the services of the
.BR mtp (4)
over an IP network using Stream Control Transmission Protocol,
.BR sctp (4),
as defined in
RFC 2960,
.[
rfc2960
.]
RFC 3309,
.[
rfc3309
.]
TS 102 144,
.[
ts 102 144
.]
and
RFC 4460.
.[
rfc4460
.]
Also, it provides a building block for redundant and distributed SS7
systems.  (See
.IR \*(lqNOTICES\*(rq ,
below.)
.PP
.B M3UA
is implemented as a
.IR STREAMS (4)
multiplexing driver.
.BR mtp "(4) or " sctp (4)
streams are linked under the
.BR M3UA
multiplexing driver using the
.IR I_LINK " or " I_PLIHK
commands of
.BR streamio (7),
to provide transport or direct MTP services
.[
q704
.]
.[
mtpv1
.]
.[
t1.111
.]
.[
jtq704
.]
to
.BR M3UA .
Once linked, the streams must be configured using
.BR m3ua_ioctl (4).
.PP
.B M3UA
provides the following stream types that are directly compatible with the
stream types provided by
.BR mtp (4).
To select one of these stream types,
.I device
can be one of the following:
.TP
.I /dev/m3ua
This device provides the
.B MTP
protocol service interface using the M-primitives of the Message Transfer Part
Interface described in
.BR mtpi (7).
This interface is suitable for use with other SS7 modules and drivers.
.TP
.IR /dev/m3ua-npi-cl ", " /dev/m3ua-npi-co
This devices provides the
.BR MTP
protocol service interface using the N-primitives of the Network Provider
Interface,
.[
NPI Revision 2.0.0
.]
.BR npi (7),
used for the Message Transfer Part as described in
.BR npi_mtp (7).
The
.BR npi (7)
interface provided on a stream opening this device is suitable for use with
the
.BR sockets (3)
library
.RB socket (3)
call
and provides the
.B MTP
interface using sockets as described in
.BR mtp_socket (7).
.I /dev/m3ua-npi-cl
provides a connectionless
.B N_CLNS
network service provider;
.I /dev/m3ua-npi-co
provides a connection-oriented
.B N_CONS
network service provider.
.TP
.IR /dev/m3ua-tpi-cl ", " /dev/m3ua-tpi-co
This device provides the
.B MTP
protocol service interface using the T-primitives of the
Transport Provider Interface,
.[
tpi revision 2.0.0 draft 2
.]
.BR tpi (7),
user for the Message Transfer Part as described in
.BR tpi_mtp (7).
The
.BR tpi (7)
interface provided on a stream opening this device is suitable for use with
the
.BR xnet (3)
or
.BR xti (3)
library
.BR t_open (3)
call and provides the
.B MTP
interface using XTI
.[
xti revision 1.0
.]
.[
xns99
.]
as described in
.BR xti_mtp (3).
.I /dev/m3ua-tpi-cl
provides a connectionless
.B T_CLNS
transport service provider;
.I /dev/m3ua-tpi-co
provides a pseudo-connection-oriented
.B T_CONS
transport service provider.
.PP
.B M3UA
provides two socket types used by the
.BR sockets (3)
library
.BR socket (3)
call.
.B M3UA
provides sockets in the protocol family
.B PF_SS7
using the protocol number
.BR SS7PROTO_M3UA .
The
.I socket_type
argument can be one of the following:
.TP
.B SOCK_RDM
Provides a connectionless datagram service with reliable delivery including
retransmission.  Message boundaaries are preserved.
This socket uses the underlying
.I /dev/m3ua-tpi-cl
transport service provider.
.TP
.B SOCK_SEQPACKET
Provides a pseudo-connection-oriented packet service with reliable delivery
including retransmission.  Message boundaries are preserved.  Messages can be
sent and received in order.
This socket uses the underlying
.I /dev/m3ua-tpi-co
transport service provider.
.PP
The following entries in
.B /etc/sock2path
provide the information necessary for
.BR socksys (4)
to use these devices:
.sp
.TS
tab(:);
llllll.
PF_SS7:SOCK_RDM:SS7PROTO_M3UA:/dev/m3ua-tpi-cl:tpi:clts
PF_SS7:SOCK_SEQPACKET:SS7PROTO_M3UA:/dev/m3ua-tpi-co:tpi:cots
.TE
.PP
For additional information, see
.BR sock2path (5).
.\"
.\"
.SH NOTICES
.PP
.B M3UA
can be configured on a system as a drop in transparent replacement for
.BR mtp (4).
In such a configuration, users opening streams or sockets as defined in
.BR mtp (4)
will be opening
.B m3ua
streams and sockets, but will be unaware of the difference.  In
.BR OpenSS7 ,
.B m3ua
is used as the primary mechanism of redundancy and distribution at the
.B MTP
level, and, when properly configured (see
.BR ss7confd (8)),
.B m3ua
functions transparently to the MTP-User.
.PP
Also, opening and binding a
.B m3ua
stream or socket may result in the opening and binding of a
.BR mtp (4)
stream on the local or remote host in direct support of the local MTP-User.
.\"
.\"
.SH IOCTLS
.PP
All interfaces support a set of management IO controls using
.BR ioctl (2s)
system calls on the
.I m3ua_stream
or
.IR m3ua_socket .
These management controls are documented in
.BR m3ua_ioctl (4).
.\"
.\"
.SH DEVICES
.PP
.IR /dev/m3ua ,
.IR /dev/m3ua-npi-cl ,
.IR /dev/m3ua-npi-co ,
.IR /dev/m3ua-tpi-cl ,
.IR /dev/m3ua-tpi-co .
.\"
.\"
.SH FILES
.PP
.BR <ss7/mtpi.h> ,
.BR <ss7/m3ua_ioctl.h> ,
.BR <sys/npi.h> ,
.BR <sys/npi_ss7.h> ,
.BR <sys/npi_mtp.h> ,
.BR <sys/tpi.h> ,
.BR <sys/tpi_ss7.h> ,
.BR <sys/tpi_mtp.h> ,
.BR <xti.h> ,
.BR <xti_mtp.h> ,
.BR <sys/socket.h> .
.\"
.\"
.SH "SEE ALSO"
.PP
.BR SS7 (4),
.BR ss7confd (8),
.BR mtpi (7),
.BR npi_mtp (7),
.BR tpi_mtp (7),
.BR xti_mtp (3),
.BR mtp_socket (7),
.BR sockets (3),
.BR xnet (3),
.BR m3ua_ioctl (4).
.\"
.\"
.SH VERSIONS
.PP
The
.B M3UA
driver is specific to the
.B OpenSS7
stack.  This is Version 2 of the interface.
.\"
.\"
.[
$LIST$
.]
.\"
.TI


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


src/drivers/m3ua.c
/*****************************************************************************

 @(#) File: src/drivers/m3ua.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/drivers/m3ua.c (" PACKAGE_ENVR ") " PACKAGE_DATE;

#include <sys/os7/compat.h>

#include "m3ua.h"
#include "m3ua_data.h"

/*
 *  M3UA MULTIPLEXOR
 *  -------------------------------------------------------------------------
 *  This is a multiplexing driver for M3UA.  When M3UA is opened by its control
 *  stream (typically a ua configuration daemon), it provides a control
 *  channel for configuration and routing.  When opened normally, by a user
 *  process, it provides an OpenSS7 MTP Provider.  OpenSS7 MTP Provider
 *  streams are I_LINKed and I_PLINKed under the multiplexor for Signalling
 *  Gateway (SG) Use by the M3UA control stream or M3UA user stream.  MTP
 *  Provider streams can be opened for Application Server (AS) use.  Transport
 *  streams (typically SCTP) are I_LINKed or I_PLINKed under the multiplexor
 *  for both SG and AS use by the UA/LM control stream.
 */

#define M3UA_DESCRIP	"M3UA STREAMS Multiplexing Driver"
#define M3UA_EXTRA	"Part of the OpenSS7 SS7 Stack for Linux Fast-STREAMS"
#define M3UA_REVISION	"OpenSS7 src/drivers/m3ua.c (" PACKAGE_ENVR ") " PACKAGE_DATE
#define M3UA_COPYRIGHT	"Copyright (c) 2008-2015  Monavacon Limited.  All Rights Reserved."
#define M3UA_DEVICE	"Part of the OpenSS7 Stack for Linux Fast-STREAMS."
#define M3UA_CONTACT	"Brian Bidulock <bidulock@openss7.org>"
#define M3UA_LICENSE	"GPL"
#define M3UA_BANNER	M3UA_DESCRIP	"\n" \
			M3UA_EXTRA	"\n" \
			M3UA_REVISION	"\n" \
			M3UA_COPYRIGHT	"\n" \
			M3UA_DEVICE	"\n" \
			M3UA_CONTACT
#define M3UA_SPLASH	M3UA_DESCRIP	"\n" \
			M3UA_REVISION

#ifdef LINUX
MODULE_AUTHOR(M3UA_CONTACT);
MODULE_DESCRIPTION(M3UA_DESCRIP);
MODULE_SUPPORTED_DEVICE(M3UA_DEVICE);
#ifdef MODULE_LICENSE
MODULE_LICENSE(M3UA_LICENSE);
#endif				/* MODULE_LICENSE */
#if defined MODULE_ALIAS
MODULE_ALIAS("streams-m3ua");
#endif
#ifdef MODULE_VERSION
MODULE_VERSION(PACKAGE_ENVR);
#endif
#endif				/* LINUX */

#define M3UA_DRV_ID	CONFIG_STREAMS_M3UA_MODID
#define M3UA_DRV_NAME	CONFIG_STREAMS_M3UA_NAME
#define M3UA_CMAJORS	CONFIG_STREAMS_M3UA_NMAJORS
#define M3UA_CMAJOR_0	CONFIG_STREAMS_M3UA_MAJOR
#define M3UA_UNITS	CONFIG_STREAMS_M3UA_NMINORS

/*
 *  =========================================================================
 *
 *  STREAMS Definitions
 *
 *  =========================================================================
 */

#define DRV_ID		M3UA_DRV_ID
#define DRV_NAME	M3UA_DRV_NAME
#define CMAJORS		M3UA_CMAJORS
#define CMAJOR_0	M3UA_CMAJOR_0
#define UNITS		M3UA_UNITS
#ifdef MODULE
#define DRV_BANNER	M3UA_BANNER
#else				/* MODULE */
#define DRV_BANNER	M3UA_SPLASH
#endif				/* MODULE */

static struct module_info m3ua_minfo = {
	DRV_ID,				/* Module ID number */
	DRV_NAME,			/* Module name */
	1,				/* Min packet size accepted *//* XXX */
	512,				/* Max packet size accepted *//* XXX */
	8 * 512,			/* Hi water mark *//* XXX */
	1 * 512				/* Lo water mark *//* XXX */
};

static struct qinit m3ua_rinit = {
	ua_rput,			/* Read put (msg from below) */
	ua_rsrv,			/* Read queue service */
	ua_open,			/* Each open */
	ua_close,			/* Last close */
	NULL,				/* Admin (not used) */
	&m3ua_minfo,			/* Information */
	NULL				/* Statistics */
};

static struct qinit m3ua_winit = {
	ua_wput,			/* Write put (msg from above) */
	ua_wsrv,			/* Write queue service */
	NULL,				/* Each open */
	NULL,				/* Last close */
	NULL,				/* Admin (not used) */
	&m3ua_minfo,			/* Information */
	NULL				/* Statistics */
};

MODULE_STATIC struct streamtab m3uainfo = {
	&m3ua_rinit,			/* Upper read queue */
	&m3ua_winit,			/* Upper write queue */
	&m3ua_rinit,			/* Lower read queue */
	&m3ua_winit			/* Lower write queue */
};

STATIC int
m3ua_init_caches(void)
{
	return (0);
}
STATIC int
m3ua_term_caches(void)
{
	return (0);
}

/*
 *  =========================================================================
 *
 *  OPEN and CLOSE
 *
 *  =========================================================================
 */

static dp_t *m3ua_opens_list = NULL;
static lp_t *m3ua_links_list = NULL;

static struct ua_driver m3ua_dinfo = {
	M3UA_CMAJOR,			/* Major device number */
	M3UA_CMINOR,			/* Number of minor devices */
	sizeof(mtp_t),			/* Private structure size */
	NULL,				/* Current control queue */
	&m3ua_lmq_u_ops,		/* LMQ User operations */
	&m3ua_ss7_u_ops,		/* SS7 User operations */
	&m3ua_opens_list,		/* Opened structures list */
	&m3ua_links_list		/* LInked structures list */
};

/*
 *  =========================================================================
 *
 *  Registration and initialization
 *
 *  =========================================================================
 */
#ifdef LINUX
/*
 *  Linux Registration
 *  -------------------------------------------------------------------------
 */

unsigned short modid = DRV_ID;

#ifndef module_param
MODULE_PARM(modid, "h");
#else
module_param(modid, ushort, 0444);
#endif
MODULE_PARM_DESC(modid, "Module ID for the INET driver. (0 for allocation.)");

major_t major = CMAJOR_0;

#ifndef module_param
MODULE_PARM(major, "h");
#else
module_param(major, uint, 0444);
#endif
MODULE_PARM_DESC(major, "Device number for the INET driver. (0 for allocation.)");

/*
 *  Linux Fast-STREAMS Registration
 *  - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 */
STATIC struct cdevsw m3ua_cdev = {
	.d_name = DRV_NAME,
	.d_str = &m3uainfo,
	.d_flag = D_MP,
	.d_fop = NULL,
	.d_mode = S_IFCHR,
	.d_kmod = THIS_MODULE,
};

STATIC int
m3ua_register_strdev(major_t major)
{
	int err;

	if ((err = register_strdev(&m3ua_cdev, major)) < 0)
		return (err);
	return (0);
}

STATIC int
m3ua_unregister_strdev(major_t major)
{
	int err;

	if ((err = unregister_strdev(&m3ua_cdev, major)) < 0)
		return (err);
	return (0);
}

MODULE_STATIC void __exit
m3uaterminate(void)
{
	int err, mindex;

	for (mindex = CMAJORS - 1; mindex >= 0; mindex--) {
		if (m3ua_majors[mindex]) {
			if ((err = m3ua_unregister_strdev(m3ua_majors[mindex])))
				cmn_err(CE_PANIC, "%s: cannot unregister major %d", DRV_NAME,
					m3ua_majors[mindex]);
			if (mindex)
				m3ua_majors[mindex] = 0;
		}
	}
	if ((err = m3ua_term_caches()))
		cmn_err(CE_WARN, "%s: could not terminate caches", DRV_NAME);
	return;
}

MODULE_STATIC int __init
m3uainit(void)
{
	int err, mindex = 0;

	cmn_err(CE_NOTE, DRV_BANNER);	/* console splash */
	if ((err = m3ua_init_caches())) {
		cmn_err(CE_WARN, "%s: could not init caches, err = %d", DRV_NAME, err);
		m3uaterminate();
		return (err);
	}
	for (mindex = 0; mindex < CMAJORS; mindex++) {
		if ((err = m3ua_register_strdev(m3ua_majors[mindex])) < 0) {
			if (mindex) {
				cmn_err(CE_WARN, "%s: could not register major %d", DRV_NAME,
					m3ua_majors[mindex]);
				continue;
			} else {
				cmn_err(CE_WARN, "%s: could not register driver, err = %d",
					DRV_NAME, err);
				m3uaterminate();
				return (err);
			}
		}
		if (m3ua_majors[mindex] == 0)
			m3ua_majors[mindex] = err;
#if 0
		LIS_DEVFLAGS(m3ua_majors[mindex]) |= LIS_MODFLG_CLONE;
#endif
		if (major == 0)
			major = m3ua_majors[0];
	}
	return (0);
}

/*
 *  Linux Kernel Module Initialization
 *  -------------------------------------------------------------------------
 */
module_init(m3uainit);
module_exit(m3uaterminate);

#endif				/* LINUX */
